File: android/app/src/main/java/com/gestorfinances/app/domain/rules/SplitCalculator.kt
package com.gestorfinances.app.domain.rules

data class SplitCalculation(
    val valid: Boolean,
    val sharesCents: List<Long> = emptyList(),
    val reason: String? = null,
)

object SplitCalculator {
    fun equal(
        totalCents: Long,
        participantCount: Int,
        payerIndex: Int,
    ): SplitCalculation {
        val validationError = validateCommon(totalCents, participantCount, payerIndex)
        if (validationError != null) return validationError

        val base = totalCents / participantCount
        val remainder = totalCents % participantCount
        val shares = MutableList(participantCount) { base }
        shares[payerIndex] += remainder
        return SplitCalculation(valid = true, sharesCents = shares)
    }

    fun percentage(
        totalCents: Long,
        basisPoints: List<Int>,
        payerIndex: Int,
    ): SplitCalculation {
        val validationError = validateCommon(totalCents, basisPoints.size, payerIndex)
        if (validationError != null) return validationError
        if (basisPoints.any { it < 0 }) {
            return SplitCalculation(valid = false, reason = "basis points must be non-negative")
        }
        val totalBasisPoints = basisPoints.sum()
        if (totalBasisPoints != 10_000) {
            return SplitCalculation(valid = false, reason = "basis points sum $totalBasisPoints != 10000")
        }

        val shares = basisPoints.map { totalCents * it / 10_000 }.toMutableList()
        val leftover = totalCents - shares.sum()
        shares[payerIndex] += leftover
        return SplitCalculation(valid = true, sharesCents = shares)
    }

    /**
     * Rescales [weightsCents] (e.g. a template's stored `split_config` line amounts) proportionally
     * to [totalCents] when the two no longer match — a recurring occurrence's confirmed amount can
     * differ from the split's own stored sum (variable-amount templates, an amount edited after
     * detection, or a NEW-detected candidate whose split came from a single source movement while
     * its amount is a group median). Floors each share, then adds the leftover to the payer's line,
     * same convention as [percentage]. Identity when the weights already sum to [totalCents].
     */
    fun rescale(
        weightsCents: List<Long>,
        totalCents: Long,
        payerIndex: Int,
    ): SplitCalculation {
        val validationError = validateCommon(totalCents, weightsCents.size, payerIndex)
        if (validationError != null) return validationError
        if (weightsCents.any { it < 0 }) {
            return SplitCalculation(valid = false, reason = "weights must be non-negative")
        }
        val weightSum = weightsCents.sum()
        val shares = if (weightSum == 0L) {
            MutableList(weightsCents.size) { 0L }
        } else {
            weightsCents.map { totalCents * it / weightSum }.toMutableList()
        }
        val leftover = totalCents - shares.sum()
        shares[payerIndex] += leftover
        return SplitCalculation(valid = true, sharesCents = shares)
    }

    fun exact(
        totalCents: Long,
        amountsCents: List<Long>,
    ): SplitCalculation {
        if (totalCents <= 0) {
            return SplitCalculation(valid = false, reason = "total must be positive")
        }
        if (amountsCents.any { it < 0 }) {
            return SplitCalculation(valid = false, reason = "amounts must be non-negative")
        }

        val sum = amountsCents.sum()
        if (sum != totalCents) {
            return SplitCalculation(valid = false, reason = "sum $sum != total $totalCents")
        }
        return SplitCalculation(valid = true, sharesCents = amountsCents)
    }

    private fun validateCommon(
        totalCents: Long,
        participantCount: Int,
        payerIndex: Int,
    ): SplitCalculation? =
        when {
            totalCents <= 0 -> SplitCalculation(valid = false, reason = "total must be positive")
            participantCount <= 0 -> SplitCalculation(valid = false, reason = "participants required")
            payerIndex !in 0 until participantCount -> {
                SplitCalculation(valid = false, reason = "payer index out of range")
            }
            else -> null
        }
}

File: windows/GestorFinances.Tests/Domain/Rules/SplitCalculator.cs
namespace GestorFinances.Domain.Rules;

public sealed record SplitCalculation(
    bool Valid,
    IReadOnlyList<long> SharesCents,
    string? Reason = null);

public static class SplitCalculator
{
    public static SplitCalculation Equal(long totalCents, int participantCount, int payerIndex)
    {
        var validationError = ValidateCommon(totalCents, participantCount, payerIndex);
        if (validationError is not null)
        {
            return validationError;
        }

        var baseShare = totalCents / participantCount;
        var remainder = totalCents % participantCount;
        var shares = Enumerable.Repeat(baseShare, participantCount).ToArray();
        shares[payerIndex] += remainder;
        return new SplitCalculation(true, shares);
    }

    public static SplitCalculation Percentage(long totalCents, IReadOnlyList<int> basisPoints, int payerIndex)
    {
        var validationError = ValidateCommon(totalCents, basisPoints.Count, payerIndex);
        if (validationError is not null)
        {
            return validationError;
        }

        if (basisPoints.Any(value => value < 0))
        {
            return new SplitCalculation(false, Array.Empty<long>(), "basis points must be non-negative");
        }

        var totalBasisPoints = basisPoints.Sum();
        if (totalBasisPoints != 10_000)
        {
            return new SplitCalculation(false, Array.Empty<long>(), $"basis points sum {totalBasisPoints} != 10000");
        }

        var shares = basisPoints.Select(value => totalCents * value / 10_000).ToArray();
        var leftover = totalCents - shares.Sum();
        shares[payerIndex] += leftover;
        return new SplitCalculation(true, shares);
    }

    /// <summary>
    /// Rescales <paramref name="weightsCents"/> (e.g. a template's stored split_config line
    /// amounts) proportionally to <paramref name="totalCents"/> when the two no longer match — a
    /// recurring occurrence's confirmed amount can differ from the split's own stored sum
    /// (variable-amount templates, an amount edited after detection, or a NEW-detected candidate
    /// whose split came from a single source movement while its amount is a group median). Floors
    /// each share, then adds the leftover to the payer's line, same convention as
    /// <see cref="Percentage"/>. Identity when the weights already sum to
    /// <paramref name="totalCents"/>. See shared/golden/template_split_rescale.json.
    /// </summary>
    public static SplitCalculation Rescale(IReadOnlyList<long> weightsCents, long totalCents, int payerIndex)
    {
        var validationError = ValidateCommon(totalCents, weightsCents.Count, payerIndex);
        if (validationError is not null)
        {
            return validationError;
        }

        if (weightsCents.Any(value => value < 0))
        {
            return new SplitCalculation(false, Array.Empty<long>(), "weights must be non-negative");
        }

        var weightSum = weightsCents.Sum();
        var shares = weightSum == 0L
            ? new long[weightsCents.Count]
            : weightsCents.Select(weight => totalCents * weight / weightSum).ToArray();
        var leftover = totalCents - shares.Sum();
        shares[payerIndex] += leftover;
        return new SplitCalculation(true, shares);
    }

    public static SplitCalculation Exact(long totalCents, IReadOnlyList<long> amountsCents)
    {
        if (totalCents <= 0)
        {
            return new SplitCalculation(false, Array.Empty<long>(), "total must be positive");
        }

        if (amountsCents.Any(value => value < 0))
        {
            return new SplitCalculation(false, Array.Empty<long>(), "amounts must be non-negative");
        }

        var sum = amountsCents.Sum();
        if (sum != totalCents)
        {
            return new SplitCalculation(false, Array.Empty<long>(), $"sum {sum} != total {totalCents}");
        }

        return new SplitCalculation(true, amountsCents.ToArray());
    }

    private static SplitCalculation? ValidateCommon(long totalCents, int participantCount, int payerIndex)
    {
        if (totalCents <= 0)
        {
            return new SplitCalculation(false, Array.Empty<long>(), "total must be positive");
        }

        if (participantCount <= 0)
        {
            return new SplitCalculation(false, Array.Empty<long>(), "participants required");
        }

        if (payerIndex < 0 || payerIndex >= participantCount)
        {
            return new SplitCalculation(false, Array.Empty<long>(), "payer index out of range");
        }

        return null;
    }
}
